internal/ansible/ansible.go

```diff
@@ -0,0 +1,221 @@
+	started := func(pid int, stdout, stderr io.ReadCloser) {
+		log.Debugf("process started: pid=%v runner_ident=%v", pid, id)
+
+		// start a goroutine to watch for event files being written to the
+		// job_events directory. When a relevant event file is detected, it gets
+		// marshaled into JSON and sent to the events channel.
+		go func() {
+			c := make(chan notify.EventInfo, 1)
+			if err := notify.Watch(filepath.Join(privateDataDir, "artifacts", id, "job_events"), c, notify.InMovedTo); err != nil {
+				log.Errorf("failed to watch private data dir: dir=%v err=%v", privateDataDir, err)
+				return
+			}
+			defer notify.Stop(c)
+
+			for e := range c {
+				log.Debugf("notify event: event=%v path=%v", e.Event(), e.Path())
+				switch e.Event() {
+				case notify.InMovedTo:
+					if strings.HasSuffix(e.Path(), ".json") {
+						data, err := os.ReadFile(e.Path())
+						if err != nil {
+							log.Errorf("cannot read file: file=%v error=%v", e.Path(), err)
+							continue
+						}
+
+						var event Event
+						if err := json.Unmarshal(data, &event); err != nil {
+							log.Errorf("cannot unmarshal data: data=%v error=%v", data, err)
+							continue
+						}
+
+						events <- event
+						log.Infof("event sent: event=%v", event)
+					}
+				}
+			}
+		}()
+
+		// start a goroutine that watches for the "status" file. When it gets
+		// written to, its contents are read, and if the status is "failed", a
+		// final "failed" event is written to the events channel. No action is
+		// taken if the status is "successful".
+		go func() {
+			c := make(chan notify.EventInfo, 1)
+			if err := notify.Watch(filepath.Join(privateDataDir, "artifacts", id, "status"), c, notify.InCloseWrite); err != nil {
+				log.Errorf("failed to watch status file: %v", err)
+			}
+			defer notify.Stop(c)
+
+			for e := range c {
+				log.Debugf("notify event: event=%v path=%v", e.Event(), e.Path())
+				switch e.Event() {
+				case notify.InCloseWrite:
+					data, err := os.ReadFile(e.Path())
+					if err != nil {
+						log.Errorf("failed to read status file: %v", err)
+						continue
+					}
+					switch string(data) {
+					case "failed":
+						// publish an "executor_on_failed" event to signal
+						// cloud connector that a run has failed.
+						events <- Event{
+							Event:     "executor_on_failed",
+							UUID:      uuid.New().String(),
+							Counter:   -1,
+							StartLine: 0,
+							EndLine:   0,
+							EventData: struct {
+								CRCDispatcherCorrelationID string `json:"crc_dispatcher_correlation_id"`
+								CRCDispatcherErrorCode     string `json:"crc_dispatcher_error_code"`
+								CRCDispatcherErrorDetails  string `json:"crc_dispatcher_error_details"`
+							}{
+								CRCDispatcherCorrelationID: correlationID,
+								CRCDispatcherErrorCode:     "UNDEFINED_ERROR",
+							},
+						}
+					case "successful":
+					default:
+						log.Errorf("unsupported status case: %v", string(data))
+					}
+				}
+			}
+		}()
```

Comment:
Same comment to this go routine. Is this go routine terminated, when related process is finished?

Reply:
My suspicion was correct. This go routine is never terminated despite related process is already finished. You  can test it, when you run trigger some ansible job and after finishing the job you can put something to the status file. e.g.:

```
echo "foo" > /var/local/lib/rhc-worker-playbook/dd2292fb-53d1-4c36-88d6-0ab2a223ec68/artifacts/dd2292fb-53d1-4c36-88d6-0ab2a223ec68/status
```

You will see in the log of worker following message:

```
Aug 15 10:55:37 fedora rhc-worker-playbook[22108]: 2024/08/15 10:55:37 notify event: event=notify.InCloseWrite path=/var/local/lib/rhc-worker-playbook/dd2292fb-53d1-4c36-88d6-0ab2a223ec68/artifacts/dd2292fb-53d1-4c36-88d6-0ab2a223ec68/status
Aug 15 10:55:37 fedora rhc-worker-playbook[22108]: 2024/08/15 10:55:37 unsupported status case: foo
```